### PR TITLE
Parallel disk write

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Summary of Changes

Fixes #

tag: @carllin

### CODE REVIEW

File: ledger/src/blockstore.rs [hunk 1]
```diff
@@ -872,12 +896,59 @@ impl Blockstore {
...
+        let mut tried = false;
+        loop {
+            info!(
+                "PUT_DATA: {}-{} {} {}",
+                slot,
+                shred.index(),
+                shred.is_data(),
+                solana_sdk::hash::hashv(&[&shred.payload]),
+            );
+            let e = Self::write_tmp(&path, &shred.payload);
```
Comment (carllin) [inline ledger/src/blockstore.rs:915]: What happens if you get two duplicate shreds? the writes to the file may race

File: ledger/src/blockstore.rs [hunk 2]
```diff
@@ -872,12 +896,59 @@ impl Blockstore {
...
     }
 
+    fn write_shred(&self, shred: &Shred) {
+        let slot = shred.slot();
+        let path = if shred.is_data() {
+            self.data_shred_path(slot, shred.index() as u64)
+        } else {
+            self.coding_shred_path(slot, shred.index() as u64)
+        };
+        let mut tried = false;
```
Comment (carllin) [inline ledger/src/blockstore.rs:906]: Can we name this something more descriptive, maybe `tried_create_slot_dir`?

File: ledger/src/blockstore.rs (hunk 1 above)
Comment (aeyakovenko) [reply ledger/src/blockstore.rs:915]: There is a check up the pipeline to prevent dups.

File: ledger/src/blockstore.rs [hunk 3]
```diff
@@ -1073,7 +1149,7 @@ impl Blockstore {
         // Commit step: commit all changes to the mutable structures at once, or none at all.
         // We don't want only a subset of these changes going through.
         let path = self.coding_shred_path(slot, shred_index);
-        Self::write_all(&path, &shred.payload)?;
+        Self::rename(&path)?;
```
Comment (carllin) [inline ledger/src/blockstore.rs:1152]: hmmm so from my understanding, `write_tmp` writes the actual shreds, but that's called from `insert_shreds`, not `insert_shreds_handle_duplicate`, and window service calls the latter: https://github.com/solana-labs/solana/blob/master/core/src/window_service.rs#L127

File: ledger/src/blockstore.rs (hunk 1 above)
Comment (carllin) [reply ledger/src/blockstore.rs:915]: hmm, but I see `write_shred` is called directly from `insert_shreds` here: https://github.com/solana-labs/solana/blob/2c3efa307d0d80873ffec2b71c423e8da5a6c3b3/ledger/src/blockstore.rs#L949, where's the duplicate check?

File: ledger/src/blockstore.rs (hunk 3 above)
Comment (aeyakovenko) [reply ledger/src/blockstore.rs:1152]: There is a replay attack check at the socket

**Comment by aeyakovenko** [comment]
Closed.  Rocks fixed wal

**Pull request closed without merging**